File: algorithms/2_medium/G/00134_gas_station/greedy.c
typedef ptrdiff_t isize;

static inline int reduce(const int* restrict arr, const isize len)
{
    int sum = 0;
    for (isize i = 0; i < len; i += 1) {
        sum += arr[i];
    }
    return sum;
}

int canCompleteCircuit(int* gas, int gas_len, int* cost, int cost_len)
{
    assert(gas_len == cost_len && gas_len > 0 && gas != NULL && cost != NULL);
    const isize len = gas_len;
    const int sum_gas = reduce(gas, len);
    const int sum_cost = reduce(cost, len);
    if (sum_gas < sum_cost) { 
        return -1;
    }

    int diff_sum = 0;
    int selected_idx = 0;

    for (isize i = 0; i < len; i += 1) {
        if (diff_sum < 0) {
            diff_sum = 0;
            selected_idx = i;
        }
        diff_sum += gas[i] - cost[i];
    }

    return (int)selected_idx;
}

File: algorithms/2_medium/G/00134_gas_station/greedy.rs
impl Solution {
    pub fn can_complete_circuit(gas: Vec<i32>, cost: Vec<i32>) -> i32 {
        assert!(gas.len() == cost.len());
        let len = gas.len();
        let sum_gas = gas.iter().map(|x| *x).sum::<i32>();
        let sum_cost = cost.iter().map(|x| *x).sum::<i32>();
        if sum_gas < sum_cost { 
            return -1;
        }

        let mut diff_sum = 0_i32;
        let mut selected_idx = 0_usize;

        for i in 0..len {
            if diff_sum < 0 {
                diff_sum = 0;
                selected_idx = i;
            }
            diff_sum += gas[i] - cost[i];
        }

        selected_idx.try_into().unwrap()
    }
}
